File: molview/src/chem/smiles.ts
import type { Atom, Bond, Molecule } from './types';
import { v3 } from './vec';
import { implicitHydrogens, isKnownElement } from './periodic';

/**
 * Разбор строки SMILES в граф молекулы.
 *
 * Поддерживается: органическое подмножество (B, C, N, O, P, S, F, Cl, Br, I),
 * атомы в квадратных скобках с зарядом и явным числом водородов,
 * кратные связи - = # :, ветвления, замыкания циклов (в том числе %nn),
 * ароматические атомы в нижнем регистре, разрыв '.'.
 *
 * Не поддерживается (сознательно, для школьного проекта не нужно):
 * стереохимия @/@@ и /\ — эти символы просто пропускаются.
 */

const ORGANIC_SUBSET = new Set(['B', 'C', 'N', 'O', 'P', 'S', 'F', 'Cl', 'Br', 'I']);
const AROMATIC_SYMBOLS = new Set(['b', 'c', 'n', 'o', 'p', 's']);

export class SmilesError extends Error {
  constructor(message: string, readonly position: number) {
    super(message);
    this.name = 'SmilesError';
  }
}

interface RawAtom {
  el: string;
  charge: number;
  aromatic: boolean;
  hCount: number;
  explicitH: boolean;
}

interface OpenRing {
  atom: number;
  order: number | null;
  aromatic: boolean;
  position: number;
}

export function parseSmiles(input: string): Molecule {
  const src = input.trim();
  if (!src) throw new SmilesError('Пустая строка', 0);

  const raw: RawAtom[] = [];
  const bonds: Bond[] = [];
  const rings = new Map<number, OpenRing>();
  const branchStack: number[] = [];

  let i = 0;
  let prev: number | null = null;
  let pendingOrder: number | null = null;
  let pendingAromatic = false;

  const fail: (msg: string) => never = (msg) => { throw new SmilesError(msg, i); };

  const linkTo = (target: number, order: number | null, aromatic: boolean, ringClosure: boolean) => {
    if (prev === null) return;
    const bothAromatic = raw[prev].aromatic && raw[target].aromatic;
    let finalOrder: number;
    let finalAromatic: boolean;
    if (order === null) {
      finalAromatic = bothAromatic;
      finalOrder = bothAromatic ? 1.5 : 1;
    } else {
      finalAromatic = aromatic;
      finalOrder = order;
    }
    bonds.push({ a: prev, b: target, order: finalOrder, aromatic: finalAromatic, ringClosure, delocalized: false });
  };

  const pushAtom = (a: RawAtom): number => {
    const id = raw.length;
    raw.push(a);
    if (prev !== null) {
      const bothAromatic = raw[prev].aromatic && a.aromatic;
      const order = pendingOrder ?? (bothAromatic ? 1.5 : 1);
      const aromatic = pendingOrder === null ? bothAromatic : pendingAromatic;
      bonds.push({ a: prev, b: id, order, aromatic, ringClosure: false, delocalized: false });
    }
    pendingOrder = null;
    pendingAromatic = false;
    prev = id;
    return id;
  };

  while (i < src.length) {
    const ch = src[i];

    // --- ветвления ---------------------------------------------------------
    if (ch === '(') {
      if (prev === null) fail('Ветвление «(» не может открывать формулу');
      branchStack.push(prev);
      i++;
      continue;
    }
    if (ch === ')') {
      const back = branchStack.pop();
      if (back === undefined) fail('Лишняя закрывающая скобка «)»');
      prev = back;
      i++;
      continue;
    }

    // --- символы связей ----------------------------------------------------
    if (ch === '-') { pendingOrder = 1; pendingAromatic = false; i++; continue; }
    if (ch === '=') { pendingOrder = 2; pendingAromatic = false; i++; continue; }
    if (ch === '#') { pendingOrder = 3; pendingAromatic = false; i++; continue; }
    if (ch === ':') { pendingOrder = 1.5; pendingAromatic = true; i++; continue; }
    if (ch === '/' || ch === '\\') { i++; continue; } // стереохимия двойной связи — игнорируем
    if (ch === '~') { pendingOrder = 1; i++; continue; }

    // --- разрыв цепи -------------------------------------------------------
    if (ch === '.') { prev = null; pendingOrder = null; i++; continue; }

    // --- замыкания циклов --------------------------------------------------
    if (ch === '%' || /[0-9]/.test(ch)) {
      let label: number;
      if (ch === '%') {
        const digits = src.slice(i + 1, i + 3);
        if (!/^[0-9]{2}$/.test(digits)) fail('После «%» должны идти две цифры номера цикла');
        label = parseInt(digits, 10);
        i += 3;
      } else {
        label = parseInt(ch, 10);
        i += 1;
      }
      if (prev === null) fail('Номер цикла не может стоять до первого атома');

      const open = rings.get(label);
      if (open === undefined) {
        rings.set(label, { atom: prev, order: pendingOrder, aromatic: pendingAromatic, position: i });
      } else {
        if (open.atom === prev) fail(`Цикл ${label} замыкается сам на себя`);
        const order = pendingOrder ?? open.order;
        const aromatic = pendingOrder !== null ? pendingAromatic : open.aromatic;
        linkTo(open.atom, order, aromatic, true);
        rings.delete(label);
      }
      pendingOrder = null;
      pendingAromatic = false;
      continue;
    }

    // --- атом в квадратных скобках ----------------------------------------
    if (ch === '[') {
      const close = src.indexOf(']', i);
      if (close < 0) fail('Не закрыта квадратная скобка «[»');
      const body = src.slice(i + 1, close);
      const atom = parseBracketAtom(body, i);
      i = close + 1;
      pushAtom(atom);
      continue;
    }

    // --- атом органического подмножества ----------------------------------
    const two = src.slice(i, i + 2);
    if (ORGANIC_SUBSET.has(two)) {
      i += 2;
      pushAtom({ el: two, charge: 0, aromatic: false, hCount: 0, explicitH: false });
      continue;
    }
    if (ORGANIC_SUBSET.has(ch)) {
      i += 1;
      pushAtom({ el: ch, charge: 0, aromatic: false, hCount: 0, explicitH: false });
      continue;
    }
    if (AROMATIC_SYMBOLS.has(ch)) {
      i += 1;
      pushAtom({ el: ch.toUpperCase(), charge: 0, aromatic: true, hCount: 0, explicitH: false });
      continue;
    }
    if (ch === '*') {
      i += 1;
      pushAtom({ el: 'C', charge: 0, aromatic: false, hCount: 0, explicitH: false });
      continue;
    }

    fail(`Непонятный символ «${ch}» в позиции ${i + 1}`);
  }

  if (branchStack.length > 0) fail('Не закрыта скобка «(»');
  if (rings.size > 0) {
    const label = [...rings.keys()][0];
    throw new SmilesError(`Цикл с номером ${label} открыт, но не замкнут`, src.length);
  }
  if (raw.length === 0) throw new SmilesError('Не найдено ни одного атома', 0);

  return buildMolecule(raw, bonds);
}

function parseBracketAtom(body: string, position: number): RawAtom {
  let p = 0;
  // изотоп — пропускаем
  while (p < body.length && /[0-9]/.test(body[p])) p++;

  let el = '';
  let aromatic = false;
  if (p < body.length && /[a-z]/.test(body[p]) && AROMATIC_SYMBOLS.has(body[p])) {
    aromatic = true;
    el = body[p].toUpperCase();
    p++;
  } else if (p < body.length && /[A-Z]/.test(body[p])) {
    el = body[p];
    p++;
    if (p < body.length && /[a-z]/.test(body[p]) && isKnownElement(el + body[p])) {
      el += body[p];
      p++;
    }
  } else {
    throw new SmilesError(`Не удалось определить элемент в «[${body}]»`, position);
  }

  // хиральность — пропускаем
  while (p < body.length && body[p] === '@') p++;

  let hCount = 0;
  if (p < body.length && body[p] === 'H') {
    p++;
    let digits = '';
    while (p < body.length && /[0-9]/.test(body[p])) digits += body[p++];
    hCount = digits === '' ? 1 : parseInt(digits, 10);
  }

  let charge = 0;
  while (p < body.length && (body[p] === '+' || body[p] === '-')) {
    const sign = body[p] === '+' ? 1 : -1;
    p++;
    let digits = '';
    while (p < body.length && /[0-9]/.test(body[p])) digits += body[p++];
    if (digits) {
      charge += sign * parseInt(digits, 10);
    } else {
      charge += sign;
      while (p < body.length && body[p] === (sign > 0 ? '+' : '-')) { charge += sign; p++; }
    }
  }

  // класс атома «:12» — пропускаем
  if (p < body.length && body[p] === ':') {
    p++;
    while (p < body.length && /[0-9]/.test(body[p])) p++;
  }

  if (p !== body.length) {
    throw new SmilesError(`Лишние символы в «[${body}]»`, position);
  }
  if (!isKnownElement(el)) {
    throw new SmilesError(`Элемент «${el}» отсутствует в справочнике программы`, position);
  }

  return { el, charge, aromatic, hCount, explicitH: true };
}

/** Развёртывание неявных водородов в явные атомы. */
function buildMolecule(raw: RawAtom[], bonds: Bond[]): Molecule {
  const bondSum = new Array(raw.length).fill(0);
  // для ароматических атомов: число связей плюс «лишняя» кратность
  // неароматических связей — см. пояснение в implicitHydrogens
  const aromaticSum = new Array(raw.length).fill(0);
  for (const b of bonds) {
    bondSum[b.a] += b.order;
    bondSum[b.b] += b.order;
    const extra = b.aromatic ? 0 : b.order - 1;
    aromaticSum[b.a] += 1 + extra;
    aromaticSum[b.b] += 1 + extra;
  }

  const atoms: Atom[] = raw.map((r, id) => ({
    id,
    el: r.el,
    charge: r.charge,
    aromatic: r.aromatic,
    hCount: r.explicitH
      ? r.hCount
      : r.aromatic
        ? implicitHydrogens(r.el, aromaticSum[id] + 1, r.charge, true)
        : implicitHydrogens(r.el, bondSum[id], r.charge),
    pos: v3(),
    placed: false,
    fromImplicitH: false,
  }));

  const allBonds: Bond[] = bonds.slice();
  const heavyCount = atoms.length;
  for (let id = 0; id < heavyCount; id++) {
    const n = atoms[id].hCount;
    for (let k = 0; k < n; k++) {
      const hid = atoms.length;
      atoms.push({
        id: hid, el: 'H', charge: 0, aromatic: false, hCount: 0,
        pos: v3(), placed: false, fromImplicitH: true,
      });
      allBonds.push({ a: id, b: hid, order: 1, aromatic: false, ringClosure: false, delocalized: false });
    }
  }

  return { atoms, bonds: allBonds };
}

/** Списки соседей для каждого атома. */
export function neighborLists(mol: Molecule): number[][] {
  const nb: number[][] = mol.atoms.map(() => []);
  for (const b of mol.bonds) {
    nb[b.a].push(b.b);
    nb[b.b].push(b.a);
  }
  return nb;
}

/** Сумма кратностей связей у атома. */
export function bondOrderSum(mol: Molecule, id: number): number {
  let s = 0;
  for (const b of mol.bonds) {
    if (b.a === id || b.b === id) s += b.order;
  }
  return s;
}

export function findBond(mol: Molecule, a: number, b: number): Bond | undefined {
  return mol.bonds.find((x) => (x.a === a && x.b === b) || (x.a === b && x.b === a));
}

File: molviewvtk/src/chem/smiles.cpp
#include "chem/smiles.h"

#include <set>
#include <utility>

#include "chem/periodic.h"

namespace chem {

namespace {

const std::set<std::string>& organicSubset() {
    static const std::set<std::string> set = {"B", "C", "N", "O", "P", "S", "F", "Cl", "Br", "I"};
    return set;
}

bool isAromaticSymbol(char c) {
    return c == 'b' || c == 'c' || c == 'n' || c == 'o' || c == 'p' || c == 's';
}

bool isDigit(char c) { return c >= '0' && c <= '9'; }
bool isLower(char c) { return c >= 'a' && c <= 'z'; }
bool isUpper(char c) { return c >= 'A' && c <= 'Z'; }
bool isSpace(char c) { return c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == '\f' || c == '\v'; }

std::string trim(const std::string& s) {
    std::size_t from = 0;
    std::size_t to = s.size();
    while (from < to && isSpace(s[from])) from++;
    while (to > from && isSpace(s[to - 1])) to--;
    return s.substr(from, to - from);
}

char upper(char c) { return isLower(c) ? static_cast<char>(c - 'a' + 'A') : c; }

/** Атом «как написан», до развёртывания неявных водородов. */
struct RawAtom {
    std::string el;
    int charge = 0;
    bool aromatic = false;
    int hCount = 0;
    bool explicitH = false;
};

/** Открытое замыкание цикла: ждёт второго появления своего номера. */
struct OpenRing {
    int atom = 0;
    bool hasOrder = false;
    double order = 0.0;
    bool aromatic = false;
};

RawAtom parseBracketAtom(const std::string& body, std::size_t position) {
    std::size_t p = 0;
    // изотоп — пропускаем
    while (p < body.size() && isDigit(body[p])) p++;

    std::string el;
    bool aromatic = false;
    if (p < body.size() && isLower(body[p]) && isAromaticSymbol(body[p])) {
        aromatic = true;
        el = std::string(1, upper(body[p]));
        p++;
    } else if (p < body.size() && isUpper(body[p])) {
        el = std::string(1, body[p]);
        p++;
        if (p < body.size() && isLower(body[p]) && isKnownElement(el + body[p])) {
            el += body[p];
            p++;
        }
    } else {
        throw SmilesError("Не удалось определить элемент в «[" + body + "]»", position);
    }

    // хиральность — пропускаем
    while (p < body.size() && body[p] == '@') p++;

    int hCount = 0;
    if (p < body.size() && body[p] == 'H') {
        p++;
        std::string digits;
        while (p < body.size() && isDigit(body[p])) digits += body[p++];
        hCount = digits.empty() ? 1 : std::stoi(digits);
    }

    int charge = 0;
    while (p < body.size() && (body[p] == '+' || body[p] == '-')) {
        const int sign = body[p] == '+' ? 1 : -1;
        const char signChar = body[p];
        p++;
        std::string digits;
        while (p < body.size() && isDigit(body[p])) digits += body[p++];
        if (!digits.empty()) {
            charge += sign * std::stoi(digits);
        } else {
            charge += sign;
            while (p < body.size() && body[p] == signChar) { charge += sign; p++; }
        }
    }

    // класс атома «:12» — пропускаем
    if (p < body.size() && body[p] == ':') {
        p++;
        while (p < body.size() && isDigit(body[p])) p++;
    }

    if (p != body.size()) {
        throw SmilesError("Лишние символы в «[" + body + "]»", position);
    }
    if (!isKnownElement(el)) {
        throw SmilesError("Элемент «" + el + "» отсутствует в справочнике программы", position);
    }

    RawAtom atom;
    atom.el = el;
    atom.charge = charge;
    atom.aromatic = aromatic;
    atom.hCount = hCount;
    atom.explicitH = true;
    return atom;
}

/** Развёртывание неявных водородов в явные атомы. */
Molecule buildMolecule(const std::vector<RawAtom>& raw, const std::vector<Bond>& bonds) {
    std::vector<double> bondSum(raw.size(), 0.0);
    // для ароматических атомов: число связей плюс «лишняя» кратность
    // неароматических связей — см. пояснение в implicitHydrogens
    std::vector<double> aromaticSum(raw.size(), 0.0);
    for (const Bond& b : bonds) {
        bondSum[b.a] += b.order;
        bondSum[b.b] += b.order;
        const double extra = b.aromatic ? 0.0 : b.order - 1.0;
        aromaticSum[b.a] += 1.0 + extra;
        aromaticSum[b.b] += 1.0 + extra;
    }

    Molecule mol;
    mol.atoms.reserve(raw.size());
    for (std::size_t id = 0; id < raw.size(); id++) {
        const RawAtom& r = raw[id];
        Atom atom;
        atom.id = static_cast<int>(id);
        atom.el = r.el;
        atom.charge = r.charge;
        atom.aromatic = r.aromatic;
        atom.hCount = r.explicitH
            ? r.hCount
            : (r.aromatic
                   ? implicitHydrogens(r.el, aromaticSum[id] + 1.0, r.charge, true)
                   : implicitHydrogens(r.el, bondSum[id], r.charge));
        mol.atoms.push_back(atom);
    }

    mol.bonds = bonds;
    const std::size_t heavyCount = mol.atoms.size();
    for (std::size_t id = 0; id < heavyCount; id++) {
        const int n = mol.atoms[id].hCount;
        for (int k = 0; k < n; k++) {
            Atom h;
            h.id = static_cast<int>(mol.atoms.size());
            h.el = "H";
            h.fromImplicitH = true;
            mol.atoms.push_back(h);

            Bond b;
            b.a = static_cast<int>(id);
            b.b = h.id;
            b.order = 1.0;
            mol.bonds.push_back(b);
        }
    }
    return mol;
}

}  // namespace

Molecule parseSmiles(const std::string& input) {
    const std::string src = trim(input);
    if (src.empty()) throw SmilesError("Пустая строка", 0);

    std::vector<RawAtom> raw;
    std::vector<Bond> bonds;
    // порядок открытия важен: в сообщении об ошибке называется первый
    // незамкнутый цикл, а не цикл с наименьшим номером
    std::vector<std::pair<int, OpenRing>> rings;
    std::vector<int> branchStack;

    std::size_t i = 0;
    int prev = -1;  // -1 означает «предыдущего атома нет» (null в оригинале)
    bool hasPendingOrder = false;
    double pendingOrder = 0.0;
    bool pendingAromatic = false;

    const auto fail = [&i](const std::string& message) {
        throw SmilesError(message, i);
    };

    // связь с уже существующим атомом (замыкание цикла)
    const auto linkTo = [&](int target, bool hasOrder, double order, bool aromatic, bool ringClosure) {
        if (prev < 0) return;
        const bool bothAromatic = raw[prev].aromatic && raw[target].aromatic;
        Bond b;
        b.a = prev;
        b.b = target;
        if (!hasOrder) {
            b.aromatic = bothAromatic;
            b.order = bothAromatic ? 1.5 : 1.0;
        } else {
            b.aromatic = aromatic;
            b.order = order;
        }
        b.ringClosure = ringClosure;
        bonds.push_back(b);
    };

    const auto pushAtom = [&](const RawAtom& a) {
        const int id = static_cast<int>(raw.size());
        raw.push_back(a);
        if (prev >= 0) {
            const bool bothAromatic = raw[prev].aromatic && a.aromatic;
            Bond b;
            b.a = prev;
            b.b = id;
            b.order = hasPendingOrder ? pendingOrder : (bothAromatic ? 1.5 : 1.0);
            b.aromatic = hasPendingOrder ? pendingAromatic : bothAromatic;
            bonds.push_back(b);
        }
        hasPendingOrder = false;
        pendingAromatic = false;
        prev = id;
    };

    while (i < src.size()) {
        const char ch = src[i];

        // --- ветвления -------------------------------------------------------
        if (ch == '(') {
            if (prev < 0) fail("Ветвление «(» не может открывать формулу");
            branchStack.push_back(prev);
            i++;
            continue;
        }
        if (ch == ')') {
            if (branchStack.empty()) fail("Лишняя закрывающая скобка «)»");
            prev = branchStack.back();
            branchStack.pop_back();
            i++;
            continue;
        }

        // --- символы связей --------------------------------------------------
        if (ch == '-') { hasPendingOrder = true; pendingOrder = 1.0; pendingAromatic = false; i++; continue; }
        if (ch == '=') { hasPendingOrder = true; pendingOrder = 2.0; pendingAromatic = false; i++; continue; }
        if (ch == '#') { hasPendingOrder = true; pendingOrder = 3.0; pendingAromatic = false; i++; continue; }
        if (ch == ':') { hasPendingOrder = true; pendingOrder = 1.5; pendingAromatic = true; i++; continue; }
        if (ch == '/' || ch == '\\') { i++; continue; }  // стереохимия двойной связи — игнорируем
        if (ch == '~') { hasPendingOrder = true; pendingOrder = 1.0; i++; continue; }

        // --- разрыв цепи -----------------------------------------------------
        if (ch == '.') { prev = -1; hasPendingOrder = false; i++; continue; }

        // --- замыкания циклов ------------------------------------------------
        if (ch == '%' || isDigit(ch)) {
            int label = 0;
            if (ch == '%') {
                const std::string digits = src.substr(i + 1, 2);
                if (digits.size() != 2 || !isDigit(digits[0]) || !isDigit(digits[1])) {
                    fail("После «%» должны идти две цифры номера цикла");
                }
                label = std::stoi(digits);
                i += 3;
            } else {
                label = ch - '0';
                i += 1;
            }
            if (prev < 0) fail("Номер цикла не может стоять до первого атома");

            auto open = rings.begin();
            while (open != rings.end() && open->first != label) ++open;
            if (open == rings.end()) {
                OpenRing r;
                r.atom = prev;
                r.hasOrder = hasPendingOrder;
                r.order = pendingOrder;
                r.aromatic = pendingAromatic;
                rings.emplace_back(label, r);
            } else {
                if (open->second.atom == prev) {
                    fail("Цикл " + std::to_string(label) + " замыкается сам на себя");
                }
                const bool hasOrder = hasPendingOrder || open->second.hasOrder;
                const double order = hasPendingOrder ? pendingOrder : open->second.order;
                const bool aromatic = hasPendingOrder ? pendingAromatic : open->second.aromatic;
                linkTo(open->second.atom, hasOrder, order, aromatic, true);
                rings.erase(open);
            }
            hasPendingOrder = false;
            pendingAromatic = false;
            continue;
        }

        // --- атом в квадратных скобках ---------------------------------------
        if (ch == '[') {
            const std::size_t close = src.find(']', i);
            if (close == std::string::npos) fail("Не закрыта квадратная скобка «[»");
            const std::string body = src.substr(i + 1, close - i - 1);
            const RawAtom atom = parseBracketAtom(body, i);
            i = close + 1;
            pushAtom(atom);
            continue;
        }

        // --- атом органического подмножества ---------------------------------
        const std::string two = src.substr(i, 2);
        if (two.size() == 2 && organicSubset().count(two) != 0) {
            i += 2;
            RawAtom a; a.el = two;
            pushAtom(a);
            continue;
        }
        if (organicSubset().count(std::string(1, ch)) != 0) {
            i += 1;
            RawAtom a; a.el = std::string(1, ch);
            pushAtom(a);
            continue;
        }
        if (isAromaticSymbol(ch)) {
            i += 1;
            RawAtom a; a.el = std::string(1, upper(ch)); a.aromatic = true;
            pushAtom(a);
            continue;
        }
        if (ch == '*') {
            i += 1;
            RawAtom a; a.el = "C";
            pushAtom(a);
            continue;
        }

        // Символ может быть многобайтовым: если забыли переключить раскладку,
        // сюда попадает кириллица. Вырезаем его целиком, иначе в сообщении
        // окажется обрубок UTF-8 и пользователь увидит «кракозябру».
        {
            std::size_t width = 1;
            const unsigned char lead = static_cast<unsigned char>(ch);
            if ((lead & 0xE0) == 0xC0) width = 2;
            else if ((lead & 0xF0) == 0xE0) width = 3;
            else if ((lead & 0xF8) == 0xF0) width = 4;
            width = std::min(width, src.size() - i);
            fail("Непонятный символ «" + src.substr(i, width) + "» в позиции "
                 + std::to_string(i + 1));
        }
    }

    if (!branchStack.empty()) fail("Не закрыта скобка «(»");
    if (!rings.empty()) {
        const int label = rings.front().first;
        throw SmilesError("Цикл с номером " + std::to_string(label) + " открыт, но не замкнут",
                          src.size());
    }
    if (raw.empty()) throw SmilesError("Не найдено ни одного атома", 0);

    return buildMolecule(raw, bonds);
}

std::vector<std::vector<int>> neighborLists(const Molecule& mol) {
    std::vector<std::vector<int>> nb(mol.atoms.size());
    for (const Bond& b : mol.bonds) {
        nb[b.a].push_back(b.b);
        nb[b.b].push_back(b.a);
    }
    return nb;
}

double bondOrderSum(const Molecule& mol, int id) {
    double s = 0;
    for (const Bond& b : mol.bonds) {
        if (b.a == id || b.b == id) s += b.order;
    }
    return s;
}

const Bond* findBond(const Molecule& mol, int a, int b) {
    for (const Bond& x : mol.bonds) {
        if ((x.a == a && x.b == b) || (x.a == b && x.b == a)) return &x;
    }
    return nullptr;
}

Bond* findBond(Molecule& mol, int a, int b) {
    for (Bond& x : mol.bonds) {
        if ((x.a == a && x.b == b) || (x.a == b && x.b == a)) return &x;
    }
    return nullptr;
}

}  // namespace chem
